drivers/i2c/i2c_pca9542a.c

```diff
@@ -0,0 +1,167 @@
+/*
+ *
+ * Copyright (c) 2020 Innoseis BV
+ * Copyright (c) 2024 tinyVision.ai Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nxp_pca9542a
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/logging/log.h>
+#include <stdint.h>
+#include <stdio.h>
+
+#define BCD2BIN(bcd) (((10 * ((bcd) >> 4)) + ((bcd) & 0x0F)))
+
+LOG_MODULE_REGISTER(pca9542a, CONFIG_I2C_LOG_LEVEL);
+
+struct pca9542a_root_config {
+	struct i2c_dt_spec i2c;
+	uint8_t nchans;
+};
+
+struct pca9542a_root_data {
+	struct k_mutex lock;
+	uint8_t selected_chan;
+};
+
+struct pca9542a_channel_config {
+	const struct device *root;
+	uint8_t chan_mask;
+};
+
+static inline struct pca9542a_root_data *get_root_data_from_channel(const struct device *dev)
+{
+	const struct pca9542a_channel_config *channel_config = dev->config;
+
+	return channel_config->root->data;
+}
+
+static inline const struct pca9542a_root_config *
+get_root_config_from_channel(const struct device *dev)
+{
+	const struct pca9542a_channel_config *channel_config = dev->config;
+
+	return channel_config->root->config;
+}
+
+static int pca9542a_configure(const struct device *dev, uint32_t dev_config)
+{
+	const struct pca9542a_root_config *cfg = get_root_config_from_channel(dev);
+
+	return i2c_configure(cfg->i2c.bus, dev_config);
+}
+
+static int pca9542a_set_channel(const struct device *dev, uint8_t select_mask)
+{
+	int res = 0;
+	struct pca9542a_root_data *data = dev->data;
+	const struct pca9542a_root_config *cfg = dev->config;
+
+	/* Only select the channel if its different from the last channel */
+	if (data->selected_chan != select_mask) {
+		res = i2c_write_dt(&cfg->i2c, &select_mask, 1);
+		if (res == 0) {
+			data->selected_chan = select_mask;
+		} else {
+			LOG_DBG("pca9542a: failed to set channel");
+		}
+	}
+	return res;
+}
+
+static int pca9542a_transfer(const struct device *dev, struct i2c_msg *msgs, uint8_t num_msgs,
+			     uint16_t addr)
+{
+	struct pca9542a_root_data *data = get_root_data_from_channel(dev);
+	const struct pca9542a_root_config *config = get_root_config_from_channel(dev);
+	const struct pca9542a_channel_config *down_cfg = dev->config;
+	int res;
+
+	res = k_mutex_lock(&data->lock, K_MSEC(5000));
+	if (res != 0) {
+		return res;
+	}
+
+	res = pca9542a_set_channel(down_cfg->root, down_cfg->chan_mask);
+	if (res != 0) {
+		goto end_trans;
+	}
+
+	res = i2c_transfer(config->i2c.bus, msgs, num_msgs, addr);
+
+end_trans:
+	k_mutex_unlock(&data->lock);
+	return res;
+}
+
+static int pca9542a_root_init(const struct device *dev)
+{
+
+	struct pca9542a_root_data *i2c_pca9542a = dev->data;
+	const struct pca9542a_root_config *config = dev->config;
+
+	printf("inside root init...\n");
```

Comment:
That looks like some debug `printf()`. Feel free to convert it to a `LOG_DBG("initializing the PCA9542A I2C multiplexer")` or even better: remove it altogether. The shell command `device list` will tell if it ran successfully.

Thanks for this cleanup!